### ISSUE TITLE
Should `expect_error(object, regexp = NA)` be considered a lint?

### ISSUE THREAD

**Issue by IndrajeetPatil** [r-lib/lintr#1312]
If I have understood the best practices correctly, if one wants to test that the function produces an output and doesn't result in an error, the way to do this is using:

```r
expect_output(object)
# or 
expect_snapshot(object)
```

But, in the wild, I see people instead using:

```r
expect_error(object, regexp = NA)
```

Should this be considered a lint?

**Comment by AshesITR** [comment]
These are different tests.
 - `expect_output()` tests for regular output (c.f. `capture.output()`).
 - `expect_snapshot()` tests for a snapshot version of output, warnings, errors and results.
 - `expect_error(..., NA)` tests that no error is generated, which is redundant in the sense that if the expression errors, the test fails automatically. Replacements could for example be `expect_silent()`, or `expect_warning()` depending on what you actually expect the expression to do.

**Comment by IndrajeetPatil** [comment]
Yeah, I think the third case is what I have in mind. 

In the context of `expect_error(..., NA)`, I am _assuming_  that what people _mean_ to use is `expect_silent(...)`; i.e., they don't expect either an error or a warning, they just expect that the function is going to work as expected.

**Comment by IndrajeetPatil** [comment]
A related lint can be using `expect_that(..., throws_error())` instead of `expect_error(...)`.
Although the latter is less frequent in the wild, at least in my experience.

**Comment by AshesITR** [comment]
The intention is hard to guess from the test code alone but it could be made into a `no_op_test_linter()` maybe...

**Comment by MichaelChirico** [comment]
TBH I am not a fan of snapshot tests

(1) they don't run on CRAN
(2) they are very fragile. every time cli or any other upstream package makes a one-character adjustment to the output formatting, your test breaks (which is why they're disabled on CRAN)

I think they are appropriate as a sort of end-to-end test, but that means they'd be used rarely.

anyway, if there is official tidyverse guidance/documentation that expect_snapshot() is preferred, I wouldn't be opposed to a PR for such a linter

**Comment by IndrajeetPatil** [comment]
Ah, no, I am not arguing the lint should recommend using snapshot tests (that was just a suggestion on my part).

As @AshesITR pointed out, the most appropriate alternative to `expect_error(object, regexp = NA)` would be to use `expect_silent(object)`.

**Comment by IndrajeetPatil** [comment]
> I think they are appropriate as a sort of end-to-end test, but that means they'd be used rarely.

Completely agree with this!

**Comment by MichaelChirico** [comment]
the expect_silent() documentation doesn't list errors, is it actually equivalent?

**Comment by AshesITR** [comment]
`expect_error(expr, NA)` is equivalent to `expr`, hence my name suggestion.

I only guessed that `expect_silent()` might be what was intended by the test. But we can never be sure about that so best we can do is highlight the No-Op.

**Comment by MichaelChirico** [comment]
I think expect_error(expr, NA) makes the intention much clearer... what's the better end user experience? how do these two fail?

test_that("fails", stop())
test_that("fails", expect_error(stop(), NA))

**Comment by IndrajeetPatil** [comment]
Hmm, I personally find it confusing to test that there is _no_ error by using `expect_error()`, but this is subjective opinion. (I wish there were an `expect_no_error()` function.)

I think given this subjectivity and the overall difficulty in inferring intent solely from `expect_error(..., NA)`, maybe this one is best left out of _lintr_, at least for now.
